diff --git a/Makefile b/Makefile
index bcbbec71a1..8b75085fa5 100644
--- a/Makefile
+++ b/Makefile
@@ -85,7 +85,7 @@ x := $(shell rm -rf meson-private meson-info meson-logs)
 endif
 
 # 1. ensure config-host.mak is up-to-date
-config-host.mak: $(SRC_PATH)/configure $(SRC_PATH)/pc-bios $(SRC_PATH)/VERSION
+config-host.mak: $(SRC_PATH)/configure $(SRC_PATH)/pc-bios $(SRC_PATH)/QEMU_VERSION
 	@echo config-host.mak is out-of-date, running configure
 	@if test -f meson-private/coredata.dat; then \
 	  ./config.status --skip-meson; \
@@ -204,7 +204,7 @@ clean: recurse-clean
 	rm -f TAGS cscope.* *.pod *~ */*~
 	rm -f fsdev/*.pod scsi/*.pod
 
-VERSION = $(shell cat $(SRC_PATH)/VERSION)
+VERSION = $(shell cat $(SRC_PATH)/QEMU_VERSION)
 
 dist: qemu-$(VERSION).tar.bz2
 
diff --git a/QEMU_VERSION b/QEMU_VERSION
new file mode 100644
index 0000000000..5214c0b8b9
--- /dev/null
+++ b/QEMU_VERSION
@@ -0,0 +1 @@
+5.2.50
diff --git a/VERSION b/VERSION
deleted file mode 100644
index 5214c0b8b9..0000000000
--- a/VERSION
+++ /dev/null
@@ -1 +0,0 @@
-5.2.50
diff --git a/meson.build b/meson.build
index 2dc66ae930..a8f8a02b3f 100644
--- a/meson.build
+++ b/meson.build
@@ -1,7 +1,7 @@
 project('qemu', ['c'], meson_version: '>=0.55.0',
         default_options: ['warning_level=1', 'c_std=gnu99', 'cpp_std=gnu++11', 'b_colorout=auto'] +
                          (meson.version().version_compare('>=0.56.0') ? [ 'b_staticpic=false' ] : []),
-        version: run_command('head', meson.source_root() / 'VERSION').stdout().strip())
+        version: run_command('head', meson.source_root() / 'QEMU_VERSION').stdout().strip())
 
 not_found = dependency('', required: false)
 if meson.version().version_compare('>=0.56.0')
